tests/providers/test_sendgrid.py

```diff
@@ -0,0 +1,624 @@
+"""
+Collection of unit and functional tests for the sendgrid library.
+Online functional tests require the following env variables:
+    NOTIFIERS_SENDGRID_API_KEY
+    NOTIFIERS_SENDGRID_TO
+The 'from' email in the online tests will be 'test@example.com'
+"""
+# pylint: disable=too-many-public-methods,no-self-use,redefined-outer-name,missing-docstring
+# pylint: disable=protected-access
+import os
+import re
+import base64
+import pytest
+from notifiers.exceptions import BadArguments
+provider = 'sendgrid'
+
+def get_basic_payload():
+    """
+    gets a payload with the minimum required arguments, can be updated
+    with test data to test specific attributes in the payload without
+    having to worry about required arguments
+    """
+    return {
+        'personalizations': [
+            {
+                'to': [
+                    {
+                        'email': 'test@example.com'
+                    }
+                ]
+            }
+        ],
+        'from': {
+            'email': 'test@example.com'
+        },
+        'subject': 'a friendly subject',
+        'content': [
+            {
+                'type': 'text/plain',
+                'value': 'my email content'
+            }
+        ],
+    }
+
+def get_attachment_payload():
+    """Gets a payload section for attachments"""
+    payload = get_basic_payload()
+    payload['attachments'] = [
+        {
+            'content': 'my content',
+            'filename': 'my file name'
+        }
+    ]
+    return payload
+
+class TestSendgridSchema:
+    """Tests just the schema validation for SG"""
+    old_environ = {}
+
+    @classmethod
+    def setup_class(cls):
+        """
+        deleting all the NOTIFIERS variables before the schema tests
+        so we can interact purely with the test data in this class
+        """
+        for each in os.environ:
+            if re.match('^NOTIFIERS.*', each):
```

Comment:
im not sure why are you doing this

Reply:
When I was running the whole test suite, I needed to set the environment variables for the live tests to run properly.  Doing that polluted the payload I was trying to verify in the schema tests because i'm calling _process_data.  I could theoretically work around that by updating my 'expected' payload to match what's in the environment, but that adds more coupling than I want to have to the way the environment variables behave (not that I don't have a little too much coupling in this suite already, but I'd still like to minimize it).  If you're cool with this approach, I can flesh out the comments a little to make it clearer.

Reply:
nah, it's my fault when not explaining, or rather documenting.
Each provider has tests that are marked as `online`. These are tests that MUST have the live envars set in order to pass (and those are configured in Travis).

You are not expected to be able to make them pass locally or via a PR. In fact, they should be skipped for PR builds.

So you should only care about the envars for your plugin, and have them added to Travis. Unfortunately, that can be done only via me currently, so if you created a "real" test account in sendgrid, you should PM me the credentials so I can add those to Travis. Then just mark each test that actually need to access the sendgrid API (and pass) with the `@pytest.mark.online` decorator.